### DISCUSSION TITLE
How to insert new values to an array

### DISCUSSION THREAD

**Discussion by charleschege** [jni-rs/jni-rs#438] [Q&A]
I have this JNI code

```rust
#[allow(non_snake_case)]
#[no_mangle]
pub extern "system" fn Java_RustLibrary_rustyArray(
    // Java environment.
    env: JNIEnv,
    // Static class which owns this method.
    _class: JClass,
) -> jarray {
    let arr_class = env.find_class("[Ljava/lang/Object;").unwrap();

    let foo = env
        .new_object_array(
            4,
            arr_class,
            JObject::from(env.new_string("UNINIT").unwrap()),
        )
        .unwrap();

    let bar = ["ONE"; 3];

    for (index, value) in bar.iter().enumerate() {
        let elem_obj = env.new_string(value).unwrap().into();
        println!("{:?}", &elem_obj);

        env.set_object_array_element(foo, index as i32, elem_obj)
            .unwrap();
    }

    foo
}
```

Which I call using he Kotlin code
```kt
@file:JvmName("RustLibrary")

external fun rustyArray(): Array<String>

val bar3 = rustyArray()

    for (value in bar3) {
        println(value)
    }
  ```
  
  It errors out :
  ```sh
  thread '<unnamed>' panicked at 'called `Result::unwrap()` on an `Err` value: Error(JavaException, State { next_error: None, backtrace: None })', src/lib.rs:166:10
note: run with `RUST_BACKTRACE=1` environment variable to display a backtrace
fatal runtime error: failed to initiate panic, error 5
```

That line of code `166` is 
```rust
env.set_object_array_element(foo, index as i32, elem_obj)
            .unwrap();
```

if I dont use the for loop in the JNI code it works though it prints "UNINT". How can I solve the issue

**Comment by argv-minus-one** [comment]
`env.set_object_array_element` has begun throwing a Java exception and returned <code>Err([jni::errors::Error::JavaException](https://docs.rs/jni/0.21.1/jni/errors/enum.Error.html#variant.JavaException))</code>. You need to handle this return value by returning from your `extern fn` by returning a null array reference, and let the Java exception be thrown. The exception's message will tell you what the actual problem is.

In a future version of this crate, we'll have a convenient wrapper for handling errors and panics, but we don't have that yet. In the mean time, you'll have to handle them yourself. See #432 for examples.

As for the reason for the error, it's that you've created not an `Object[]` but an `Object[][]`. To fix this, change `arr_class` to `java/lang/Object` instead of `[Ljava/lang/Object;`.
